src/lib.rs

```diff
@@ -4,4 +4,161 @@
 //! Future implementations will support the [ESP32-WROOM-DA](https://www.espressif.com/sites/default/files/documentation/esp32-wroom-da_datasheet_en.pdf) module.
 //! 
-//! NOTE This crate is still under active development. This API will remain volatile until 1.0.0
\ No newline at end of file
+//! NOTE This crate is still under active development. This API will remain volatile until 1.0.0
+
+
+// This is just a placeholder for now. We will need to implement a socket struct or
+// use an existing crate
+type Socket = [u8; 1];
+
+// TODO: Find a way to share these between NinaCommandHandlers
+const START_CLIENT_TCP: u8 = 0x2du8;
+const GET_FW_VERSION: u8 = 0x37u8;
+
+pub struct Wifi<C: NinaCommandHandler<I>> {
+  command_handler: C,
+}
+
+pub struct FirmwareVersion {
+    major: u8,
+    minor: u8,
+    patch: u8
+}
+
+impl FirmwareVersion {
+    fn new(version: [u8; 5]) -> FirmwareVersion {
+        FirmwareVersion::parse(version)
+    }
+
+    // Takes in 5 bytes (e.g. 1.7.4) and returns a FirmwareVersion instance
+    fn parse(version: [u8; 5]) -> FirmwareVersion {
+        // TODO: real implementation
+        FirmwareVersion {
+            major: 1,
+            minor: 7,
+            patch: 4
+        }
+    }
+}
+
+impl<I: IoInterface, C: NinaCommandHandler<I>> Wifi<I, C> {
+    // fn connect(&self) -> Result<T> {
+    //     self.command_handler.start_client_tcp()
+    // }
+
+    fn get_firmware_version(&self) -> Result<FirmwareVersion, Error> {
+      self.command_handler.get_fw_version()
+    }
+}
+
+impl SpiCommandHandler<I> {
+    fn send_command(command: u8, parameters: [u8; 5]) -> Result<FirmwareVersion, Error> {
+        Ok(FirmwareVersion::new([1,0,7,0,4]))
+      }
+}
+
+impl<I> NinaCommandHandler<I> for SpiCommandHandler<I> {
+
+    type Error = dyn std::error::Error;
+
+    fn start_client_tcp(&self, socket: Socket) -> Result<FirmwareVersion, Error> {
+        // TODO: implement a trait interface and set of structs for different parameter sets, e.g. SocketType
+        SpiCommandHandler::send_command(START_CLIENT_TCP, [socket])
+    }
+
+    fn get_fw_version(&self) -> Result<FirmwareVersion, Error> {
+        SpiCommandHandler::send_command(GET_FW_VERSION, [])
+    }
+}
+
+trait NinaCommandHandler<I> {
+  const START_CLIENT_TCP: u8;
+  const GET_FW_VERSION: u8;
+
+  type Error;
+
+  fn start_client_tcp(&self) -> Result<FirmwareVersion, Error>;
+
+  fn get_fw_version(&self) -> Result<FirmwareVersion, Error>;
+}
+
+struct SpiCommandHandler<I: IoInterface> {
+    io_interface: I
+}
+
+trait IoInterface {
+
+  fn esp_select(&mut self);
+
+  fn esp_deselect(&mut self);
+
+  fn get_esp_ready(&self);
+
+  fn get_esp_ack(&self);
+
+  fn wait_for_esp_ready(&self);
+
+  fn wait_for_esp_ack(&self);
+
+  fn wait_for_esp_select(&mut self);
+  
+}
+
+struct IoInterfaceImpl {
+  esp32_pins: Esp32Pins
```

Comment:
Let's refer to the target wifi board as esp, so this would become `esp_pins`. This will allow us to be consistent and hold to our intention to not rename the lower level function names that the nina-fw has on for ESP32 chip.

Reply:
Absolutely. Updated in 54c058e20f0ee9145c537322649197c9bf3df2b4.